webofneeds/won-node/src/main/java/won/node/camel/processor/facet/groupFacet/SendMessageFromNodeGroupFacetImpl.java

```diff
@@ -31,51 +31,84 @@
-    }
-    for (final Connection conToSendTo : consInGroup) {
-      try {
-        if (!conToSendTo.equals(conOfIncomingMessage)) {
-          if (logger.isDebugEnabled()) {
-            logger.debug("sending group message {} from need {} to need {}", new Object[]{wonMessage.getMessageURI(), wonMessage.getSenderNeedURI(), conToSendTo.getRemoteNeedURI()});
-          }
-          URI forwardedMessageURI = wonNodeInformationService.generateEventURI(wonMessage.getReceiverNodeURI());
-          URI remoteWonNodeUri = WonLinkedDataUtils.getWonNodeURIForNeedOrConnectionURI(conOfIncomingMessage.getRemoteConnectionURI(),
-            linkedDataSource);
-          WonMessage newWonMessage = WonMessageBuilder.forwardReceivedNodeToNodeMessageAsNodeToNodeMessage(
-            forwardedMessageURI, wonMessage,
-                  conToSendTo.getConnectionURI(), conToSendTo.getNeedURI(), wonMessage.getReceiverNodeURI(),
-                  conToSendTo.getRemoteConnectionURI(), conToSendTo.getRemoteNeedURI(), remoteWonNodeUri);
-          sendSystemMessage(newWonMessage);
+        //now check if we processed the message earlier
+
+        if (messageEventRepository.existEarlierMessageWithSameInnermostMessageURIAndReceiverNeedURI(wonMessage.getMessageURI())) {
+            if (logger.isDebugEnabled()) {
+                URI innermostMessageURI = wonMessage.getInnermostMessageURI();
+                URI groupUri = wonMessage.getReceiverNeedURI();
+                logger.debug("suppressing message {} " +
+                                "as its innermost message is {} which has already " +
+                                "been processed by group {}",
+                        new Object[]{wonMessage.getMessageURI(), innermostMessageURI, groupUri});
+            }
+            return;
+        }
+
+        final Connection conOfIncomingMessage = connectionRepository.findByConnectionURI(wonMessage.getReceiverURI()).get(0);
+
+        final List<Connection> consInGroup = connectionRepository.findByNeedURIAndStateAndTypeURI(conOfIncomingMessage.getNeedURI(),
+                ConnectionState.CONNECTED, FacetType.GroupFacet.getURI());
+
+        if (consInGroup == null || consInGroup.size() < 2) return;
+        if (logger.isDebugEnabled()) {
+            logger.debug("processing message {} received from need {} in group {} - preparing to send it to {} group members (text message: '{}'}", new Object[]{wonMessage.getMessageURI(), wonMessage.getSenderNeedURI(), wonMessage.getReceiverNeedURI(), consInGroup.size() - 1, WonRdfUtils.MessageUtils.getTextMessage(wonMessage)});
+        }
+        for (final Connection conToSendTo : consInGroup) {
+            try {
+                if (!conToSendTo.equals(conOfIncomingMessage)) {
+                    if (messageEventRepository.isReceivedSameInnermostMessageFromSender(wonMessage.getMessageURI(), conToSendTo.getRemoteNeedURI())){
```

Comment:
i would suggest to get rid of this method that checks before sending if a message was already received at the receiver side. This is kind of duplicate check since its anyway checked when messages are received. I would vote in favor for complexity reduction here instead of posible performance gain here

Reply:
Ok, as discussed, doesnt do much anyway.